lib/impure/db_sqlite.nim

```diff
@@ -716,8 +719,67 @@ proc setEncoding*(connection: DbConn, encoding: string): bool {.
 proc finalize*(sqlPrepared:SqlPrepared){.discardable.} = 
   discard finalize(sqlPrepared.PStmt)
 
+template dbBindParamError*(paramIdx: int, val: untyped) =
+  ## Raises a `DbError` exception.
+  var e: ref DbError
+  new(e)
+  e.msg = "error binding param in position " & $paramIdx 
+  raise e
+
+proc bindParam*(ps: SqlPrepared, paramIdx: int, val: int32) {.since: (1, 3).} =
+  ## Binds a int32  to the specified paramIndex.
+  if SQLITE_OK != bind_int(ps.PStmt, paramIdx.int32, val):
+    dbBindParamError(paramIdx, val)
+
+
+proc bindParam*(ps: SqlPrepared, paramIdx: int, val: int) {.since: (1, 3).} =
+  ## Binds a int  to the specified paramIndex.
+  when sizeof(int) == 8:
+    if SQLITE_OK != bind_int64(ps.PStmt, paramIdx.int32, val.int64):
+      dbBindParamError(paramIdx, val)
+  else:
+    if SQLITE_OK != bind_int(ps.PStmt, paramIdx.int32, val.int32):
+      dbBindParamError(paramIdx, val)
+
+proc bindParam*(ps: SqlPrepared, paramIdx: int, val: int64) {.since: (1, 3).} =
+  ## Binds a int64  to the specified paramIndex.
+  if SQLITE_OK != bind_int64(ps.PStmt, paramIdx.int32, val):
+    dbBindParamError(paramIdx, val)
+
+proc bindParam*(ps: SqlPrepared, paramIdx: int, val: float64) {.since: (1, 3).} =
+  ## Binds a 64bit float to the specified paramIndex.
+  if SQLITE_OK != bind_double(ps.PStmt, paramIdx.int32, val):
+    dbBindParamError(paramIdx, val)
+
+proc bindNull*(ps: SqlPrepared, paramIdx: int) {.since: (1, 3).} =
+  ## Sets the bindparam at the specified paramIndex to null 
+  ## (default behaviour by sqlite).
+  if SQLITE_OK != bind_null(ps.PStmt, paramIdx.int32):
+    dbBindParamError(paramIdx, val)
+
+proc bindParam*(ps: SqlPrepared, paramIdx: int, val: string) {.since: (1, 3).} =
+  ## Binds a string to the specified paramIndex.
+  if SQLITE_OK != bind_text(ps.PStmt, paramIdx.int32, val.cstring, -1.int32, SQLITE_STATIC):
+    dbBindParamError(paramIdx, val)
+
+proc bindParam*(ps: SqlPrepared, paramIdx: int,val: openArray[byte]) {.since: (1, 3).} =
+  ## binds a blob to the specified paramIndex.
+  let len = val.len
+  if SQLITE_OK != bind_blob(ps.PStmt, paramIdx.int32, val[0].unsafeAddr, len.int32, SQLITE_STATIC):
```

Comment:
Is `SQLITE_STATIC` the way to go here? or should be use `SQLITE_TRANSIENT`?

Also, can you follow the general style in this module and compare with `SQLITE_OK` on the right hand side?

Reply:
ok, I change it to right hand, SQLITE_TRANSIENT will cause sqlite copy the value.

Reply:
Yea but if someone modifies the array right after they call `bindParam`, then would it keep using the old array as the caller would expects?

Reply:
I have not go that further, if someone modify the param, the prepared can store right value without through bind_* ? if so I assuming developer know what they do when they change the bound value

Reply:
Then maybe add a warning stating that the value will not be copied?

An another factor that worries me is when the GC collects the memory, for example:

```nim
proc bindFoo(ps: SqlPrepared, idx: int) =
  var s = "foo"
  ps.bindParam(idx, s)

  # Once control flow exit the proc, `s` might be collected and
  # the data will be invalid!
```

No copy will only be safe if you're in control of the data for the entire period of bind->execution->reset.

Reply:
hmm, for now I think just give comment note, maybe implement param pass by value version, or type that has destructor, and also a template runnning entire period